interfaces.go

```diff
@@ -0,0 +1,223 @@
+	Add(key string, item interface{})
+
+	// Done indicates that the caller finished working on items represented by a unique key
+	// if it has been added again while it was being processed, it will be re-added to the queue for re-processing
+	Done(key string)
+}
+
+// BatchQueue represents work queue that process EndpointSamples in the order in which they were added,
+// it also supports batching - items are grouped by a key and could be retrieved as a package
+type endPointSampleBatchQueue interface {
+	// Get retrieves the next batch of collected EndpointSamples along with the unique key
+	// A caller must execute the corresponding Done() method once it has finished its work
+	Get() (key string, items []*EndpointSample)
+
+	// Add adds the given EndpointSample under the given key to the queue
+	Add(key string, item *EndpointSample)
+
+	// Done indicates that the caller finished working on the batch of EndpointSamples represented by a unique key
+	// if it has been added again while it was being processed, it will be re-added to the queue for re-processing
+	Done(key string)
+}
+
+// endpointSampleBatchQueue implements endPointSampleBatchQueue
+type endpointSampleBatchQueue struct {
+	BatchQueue
+}
+
+// Get retrieves the next batch of collected EndpointSamples along with the unique key
+// A caller must execute the corresponding Done() method once it has finished its work
+func (q *endpointSampleBatchQueue) Get() (key string, items []*EndpointSample) {
+	key, rawEndpointSample := q.BatchQueue.Get()
+	endpointSamples := make([]*EndpointSample, len(rawEndpointSample))
+	for i, r := range rawEndpointSample {
+		endpointSamples[i] = r.(*EndpointSample)
+	}
+	return key, endpointSamples
+}
+
+// Add adds the given EndpointSample under the given key to the queue
+func (q *endpointSampleBatchQueue) Add(key string, item *EndpointSample) {
+	q.BatchQueue.Add(key, item)
+}
+
+// Done indicates that the caller finished working on the batch of EndpointSamples represented by a unique key
+// if it has been added again while it was being processed, it will be re-added to the queue for re-processing
+func (q *endpointSampleBatchQueue) Done(key string) {
+	q.BatchQueue.Done(key)
+}
+
+// newEndPointSampleBatchQueue creates a strongly typed batch queue from the delegate
+// the returned queue implements endPointSampleBatchQueue interface
+func newEndPointSampleBatchQueue(delegate BatchQueue) endPointSampleBatchQueue {
+	return &endpointSampleBatchQueue{BatchQueue: delegate}
+}
+
+// EndpointSample represents a sample collected for an endpoint derived from a proxied request.
+// it holds:
+//  - namespace, service and url to uniquely identify the request
+//  - an optional err returned from the proxy
+type EndpointSample struct {
+	namespace string
+	service   string
+	url       *url.URL
+	err       error
+}
+
+// WeightedEndpointStatus represents the current status of the given endpoint based on the collected samples.
```

Comment:
in my own words: WeightedEndpointStatus maintains a ring buffer the last n samples and its computer weight for an endpoint.